src/ray/util/event.cc

```diff
@@ -0,0 +1,106 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/event.h"
+
+namespace ray {
+
+EventManager &EventManager::Instance() {
+  static EventManager INSTANCE;
+  return INSTANCE;
+}
+
+bool EventManager::IsEmpty() { return reporter_map_.empty(); }
+
+void EventManager::Publish(rpc::Event &event) {
+  for (const auto &element : reporter_map_) {
+    (element.second)->Report(event);
+  }
+}
+
+void EventManager::AddReporter(std::shared_ptr<LogBasedEventReporter> reporter) {
+  reporter_map_.emplace(reporter->GetReporterKey(), reporter);
+}
+
+void EventManager::ClearReporters() { reporter_map_.clear(); }
+
+thread_local std::unique_ptr<RayEventContext> RayEventContext::instance_ = nullptr;
+
+RayEventContext &RayEventContext::Instance() {
```

Comment:
const?

Reply:
I delete the copy constructor and the equal operator, I think it also makes sense?
```C++
RayEventContext(const RayEventContext &event_context) = delete;
const RayEventContext &operator=(const RayEventContext &event_context) = delete;
```